Add support for Iv2-prefixed Client IDs into Github App Credential

stevensonmichel opened the pull request:
As Github encourages more the use of Client IDs, one can create a Github App Installation Lookup App Credential using Github Client IDs starting with Iv1. However, using Iv2 will throw an error.

Fixes: as Github is providing Client IDs starting with Iv2, Credentials should be able to be created with it. The changes in this PR adds support for Iv2, enabling Credentials to be created using them.

arrestle commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -139,8 +139,15 @@ def _is_intish(app_id_candidate: str | int) -> bool:
-    client_id_prefix = 'Iv1.'
-    if not client_id_candidate.startswith(client_id_prefix):
+    # Support both Iv1 and Iv2 client ID prefixes
+    valid_prefixes = ('Iv1.', 'Iv2.')
```
I would recommend allowing "Iv#." Since you don't want to break again if they come up with Iv3. next year.

stevensonmichel replied:
Sure

arrestle reviewed (dismissed):
Great start! Python has https://docs.python.org/3/library/re.html or https://www.w3schools.com/python/python_regex.asp

webknjaz commented on tests/github_app_test.py:
```diff
@@ -279,8 +297,22 @@ def token(self: '_FakeAppInstallationAuth') -> str:
+        'app-id-str',
+        'client-id-iv1',
+        'client-id-iv2',
+        'client-id-iv23',
```
These IDs don't explain what the respective tests are supposed to check. Could you make it clearer?

webknjaz replied:
Additionally, with the list of tests becoming longer, it's probably a good idea to move IDs into `pytest.param()` definitions and keep them coupled. It's okay for 2-3 tests but it'd difficult to maintain two coupled iterables when they are long.

stevensonmichel replied:
Fixed

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -132,28 +132,23 @@ class EmptyKwargs(TypedDict):
 
 def _is_client_id(client_id_candidate: str) -> bool:
-    client_id_prefix = 'Iv1.'
-    if not client_id_candidate.startswith(client_id_prefix):
-        return False
-
-    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
-
-    if len(client_id_trailer) != GH_CLIENT_ID_TRAILER_LENGTH:
-        return False
```
It might be a good idea to have two validation paths — for v1 and for v2. Then, v1 could remain stricter.

stevensonmichel replied:
Alright! Resolved

webknjaz commented on tests/github_app_test.py:
```diff
@@ -165,7 +173,17 @@ class AppInstallIds(TypedDict):
+            '^Expected GitHub App Installation ID to be an integer '
+            "but got 'invalid'$",
+            id='gh-app-invalid-install-id-with-client-id-iv2',
             marks=github_app_jwt_client_id_unsupported,
```
Is this even needed given that the test checks the install ID and doesn't care what's in the app/client ID? The additiona seems redundant.

webknjaz commented on tests/github_app_test.py:
```diff
@@ -165,7 +173,17 @@ class AppInstallIds(TypedDict):
             "but got 'invalid'$",
-            id='gh-app-invalid-install-id-with-client-id',
+            id='gh-app-invalid-install-id-with-client-id-iv1',
+            marks=github_app_jwt_client_id_unsupported,
```
Following the linked issue chain, it seems that it might work now. It would be worth checking if anything changed before makring another test with xfail.

webknjaz replied:
@stevensonmichel I'm still waiting for the results. Marking the thread as resolved does not mean that this is resolved. It's not.

webknjaz commented:
@stevensonmichel FYI adding merge commits into PR branches creates foxtrots, which is why using rebase is preferred.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +158,31 @@ def _is_client_id(client_id_candidate: str) -> bool:
+    if not client_id_candidate.startswith('Iv1'):
+        return False
+    if len(client_id_candidate) <= 3:
+        return True
```
Why is this true?

stevensonmichel replied:
This is just to prevent and indexError crash. A suitable example would be a client id with only 3 characters

webknjaz replied:
Turned out this is confusing to parse because past the first guard expression, all values of `client_id_candidate` would start with `Iv1` which is 3 exactly. And checking for less than 3 is counterintuitive.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -132,18 +132,20 @@ class EmptyKwargs(TypedDict):
 GH_CLIENT_ID_TRAILER_LENGTH = 16
 HEXADECIMAL_BASE = 16
+GH_CLIENT_ID_MIN_LENGTH = 19
+GH_CLIENT_ID_MAX_LENGTH = 22
```
Let's try keeping coupled values together:
```suggestion
GH_CLIENT_ID_LENGTH_BOUNDS = range(19, 23)  # min 19, max 22
```

stevensonmichel replied:
Fixed

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +158,31 @@ def _is_client_id(client_id_candidate: str) -> bool:
+        client_id_candidate.startswith('Iv')
+        and GH_CLIENT_ID_MIN_LENGTH
+        <= len(client_id_candidate)
+        <= GH_CLIENT_ID_MAX_LENGTH
```
```suggestion
        and len(client_id_candidate) in GH_CLIENT_ID_LENGTH_BOUNDS
```

stevensonmichel replied:
Fixed

webknjaz reviewed (commented):
I feel like keeping a range in the same object is a bit more readable.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
+        return False
+    if len(client_id_candidate) <= 3:
+        return True
+    return not client_id_candidate[3].isdigit()
```
```suggestion
    known_v1_prefix = 'Iv1'

    if not client_id_candidate.startswith(known_v1_prefix):
        return False

    client_id_candidate_trailer = client_id_candidate.removeprefix(known_v1_prefix)

    if not client_id_candidate_trailer:  # empty trailer, no digits to check
        return True

    return not client_id_candidate_trailer[0].isdigit()
```

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
     return True
 
 
+def _is_iv1_prefix(client_id_candidate: str) -> bool:
```
Looking closer, I don't understand why you don't check for `Iv1.`

arrestle replied:
because github switched to access tokens with a prefix of Iv2 with possible extensions to other prefixes... I'm not sure checking for ANY prefix makes sense any more. see https://docs.github.com/en/rest/apps/oauth-applications?apiVersion=2022-11-28

webknjaz replied:
Not sure what that doc demonstrates. We already know that any validation is hunch-based. I'm actually starting to think that validation would be more useful in a form of a non-blocking hint rather than a thing that tries to predict whether input is valid. At least, for the inputs where GH doesn't document their exact shape.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
+    # Iv2+: just check length (19-22 chars)
+
+    if _is_iv1_prefix(client_id_candidate):
+        return _is_iv1_client_id(client_id_candidate)
```
Why can't this just be
```suggestion
    if _is_iv1_client_id(client_id_candidate):
        return True
```
?

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -132,18 +132,19 @@ class EmptyKwargs(TypedDict):
         return False
 
-    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
+    client_id_trailer = client_id_candidate.removeprefix('Iv1.')
```
No need to duplicate literals in multiple places. Revert this to the original forms.
```suggestion
    client_id_prefix = 'Iv1.'
    if not client_id_candidate.startswith(client_id_prefix):
        return False

    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
```

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
+        return True
+    return not client_id_candidate[3].isdigit()
+
+
```
With other observations, this becomes unnecessary.
```suggestion
```

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -132,18 +132,19 @@ class EmptyKwargs(TypedDict):
 
 GH_CLIENT_ID_TRAILER_LENGTH = 16
 HEXADECIMAL_BASE = 16
+GH_CLIENT_ID_LENGTH_BOUNDS = (19, 22)
```
Let's unbreak this by applying the original suggestion I made in https://github.com/ansible/awx-plugins/pull/139#discussion_r2473386861
```suggestion
GH_CLIENT_ID_LENGTH_BOUNDS = range(19, 23)  # min 19, max 22
```

`thing in (1, 42)` will only be true if `thing == 1 or thing == 42`. Using the range allows anything in-between. Just take into account that the second argument is not included in the range. So `3 in range(1, 42) is True` and `41 in range(1, 42) is True`, but `42 in range(1, 42) is False`.

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv23toolongtobevalidids'$",
+            id='gh-app-id-iv2-too-long',
```
To catch bugs like the one I found in, https://github.com/ansible/awx-plugins/pull/139/files?new_files_changed=true#r2478577419, add tests for the lengths matching the boundaries — with 18/19/22/23 chars exactly.

arrestle replied:
I would remove Iv1 and Iv2 - possibly just `^Iv[1-9]\d*\.[a-zA-Z0-9]+$` or use the endpoint 

`curl -L \
  -X POST \
  -H "Accept: application/vnd.github+json" \
  -u "<YOUR_CLIENT_ID>:<YOUR_CLIENT_SECRET>" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  "https://api.github.com/applications/<CLIENT_ID>/token" \
  -d '{"access_token":"<TOKEN_TO_VALIDATE>"}'`
  
See https://docs.github.com/en/rest/apps/oauth-applications?apiVersion=2022-11-28#check-a-token or https://docs.github.com/en/rest/apps/oauth-applications?apiVersion=2022-11-28

webknjaz replied:
Well, since there's already network I/O for retrieving the token, it will tell us if the creds are valid right in there. But it probably won't specify whether it's the client id or the token that's wrong.
If a regex was to be used, it'd probably need to differentiate between v1 and the rest. Though, I'm not a fan of regexes for such simple checks since they're more difficult to maintain. The coverage report won't tell you that you messed us the range boundaries in your regex, for example and it won't hint you how to cover unobvious cases.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on tests/github_app_test.py:
```diff
@@ -273,19 +346,36 @@ def token(self: '_FakeAppInstallationAuth') -> str:
-    ids=('install-id-int', 'install-id-str'),
+    (
+        pytest.param(TEST_INSTALLATION_ID, id='install-id-int'),
+        pytest.param(str(TEST_INSTALLATION_ID), id='install-id-str-numeric'),
```
I wouldn't touch unrelated places. This will end up extending the scope of the PR infinitely. It's a bad tone to smash together different types of changes (like refactoring/formatting/functional). Pick one and stick with it so that the PR is atomic and only contains changes that contribute to a single declared goal.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +157,28 @@ def _is_client_id(client_id_candidate: str) -> bool:
+    if not client_id_candidate.startswith(known_v1_prefix):
+        return False
+
+    client_id_candidate_trailer = client_id_candidate.removeprefix(
```
This should unbreak the CI
```suggestion
```

(*I think* it's GitHub's new experimental review interface that might've corrupted the patch)

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -132,18 +132,19 @@ class EmptyKwargs(TypedDict):
         return False
 
-    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
+    client_id_trailer = client_id_candidate.removeprefix('Iv1.')
```
```suggestion
    client_id_prefix = 'Iv1.'
    if not client_id_candidate.startswith(client_id_prefix):
        return False

    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
```

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
-            id='gh-app-id-client-id-not-enough-chars',
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1'$",
```
```suggestion
            "but got 'Iv1'$",
```

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
-            id='gh-app-id-client-id-broken-hex',
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1.bbb'$",
```
```suggestion
            "but got 'Iv1.bbb'$",
```

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
-            id='gh-app-id-client-id-too-many-chars',
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1aaaaaaaaaaaaaaaa'$",
```
```suggestion
            "but got 'Iv1aaaaaaaaaaaaaaaa'$",
```

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1.gggggggggggggggg'$",
```
```suggestion
            "but got 'Iv1.gggggggggggggggg'$",
```

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1.aaaaaaaaaaaaaaaaa'$",
```
```suggestion
            "but got 'Iv1.aaaaaaaaaaaaaaaaa'$",
```

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv'$",
```
```suggestion
            "but got 'Iv'$",
```

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv2.bbb'$",
```
```suggestion
            "but got 'Iv2.bbb'$",
```

webknjaz commented on tests/github_app_test.py:
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv23toolongtobevalidids'$",
```
```suggestion
            "but got 'Iv23toolongtobevalidids'$",
```

arrestle reviewed (dismissed):
I would avoid relying on string comparison and use [this github endpoint](https://docs.github.com/en/rest/authentication/authenticating-to-the-rest-api?apiVersion=2022-11-28) to validate a token, if indeed this is even a requirement. 

The github token code may be in flux, and while currently seems to be Iv[1|2].[n+] future versions may not even follow that pattern.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -132,28 +133,48 @@ class EmptyKwargs(TypedDict):
+GH_CLIENT_ID_LENGTH_BOUNDS = range(
+    GH_CLIENT_ID_MIN_LENGTH,
+    GH_CLIENT_ID_MAX_LENGTH + 1,
+)  # min 19, max 22
```
Since you've reintroduced explicit constants, this comment is no longer necessary. In fact, there is harm in it potentially going out of sync and becoming confusing.
```suggestion
)
```

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -132,28 +133,48 @@ class EmptyKwargs(TypedDict):
     return isinstance(app_id_candidate, int) or app_id_candidate.isdigit()
 
 
 def _is_client_id(client_id_candidate: str) -> bool:
```
Plz use a helper function instead of increasing indentation.

webknjaz commented on src/awx_plugins/credentials/github_app.py:
```diff
@@ -156,6 +163,28 @@ def _is_client_id(client_id_candidate: str) -> bool:
+
+    # Check for invalid Iv1 format: Iv1 followed by non-digit
+    # (not dot, not digit). Examples: Iv1aaa, Iv1xxx → INVALID
+    if re.match(r'^Iv1[^\d]', client_id_candidate):
```
At this point, it should never be a v1 since `_is_iv1_client_id()` would be `False` here. But if we want this additional check, it belongs in that helper and shouldn't be in disconnected context.

stevensonmichel replied:
Now, I see. As it is to validate cases like Iv10.... , I think my comment "Check for Iv1 format" is misleading. I will change that.

stevensonmichel replied:
Also, I tested PyGithub with client Ids, and it validated it return the token ghs_...

webknjaz replied:
> I tested PyGithub with client Ids, and it validated it return the token ghs_...

@stevensonmichel good to know. I see that you've already taken care of the `xfail` markers so I'm marking this thread as resolved.

webknjaz commented:
In general, I agree with @arrestle in that we don't know anything about these tokens and might as well just drop any shape/length validation.

P.S. I'm still interested in learning if the underlying libraries are finally able to accept client IDs, though, and dropping `xfails` so it's actually tested somehow.

webknjaz commented on tests/github_app_test.py:
```diff
@@ -97,90 +93,9 @@ class AppInstallIds(TypedDict):
-        )
+# NOTE: Format validation tests removed - we no longer validate
+# Client ID formats upfront. GitHub's API is the ultimate validator.
+# Invalid credentials will be rejected by the API with appropriate errors.
```
Make sure not to keep comments that aren't useful in files.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on tests/github_app_test.py:
```diff
@@ -29,12 +29,8 @@
-    'Ref: https://github.com/PyGithub/PyGithub/issues/3213',
-)
+# Client IDs are now supported - no longer using xfail marker
+# All Client ID tests should pass with the no-validation approach
```
This detached statement is not needed. It doesn't add clarity, just confusion.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on tests/github_app_test.py:
rename this to
```suggestion
    'application_or_client_id',
```

webknjaz replied:
Marking this thread as resolved.

webknjaz commented:
@stevensonmichel macos will stop failing once you rebase.

stevensonmichel committed:
Support GitHub App Client IDs (Iv2+ format) in credentials

Remove upfront validation of app_or_client_id and rely on GitHub API
to validate credentials. This allows the plugin to support newer GitHub
Client ID formats (e.g., Iv23likIfIXeZTb5GBUU) without breaking when
GitHub introduces new formats in the future.

Changes:
- Removed validation functions that checked Client ID format/length
- Updated extract_github_app_install_token to accept any input
- Removed obsolete test cases for format validation
- Added tests for various Client ID formats including Iv23 prefix
- Updated parameter names to application_or_client_id for consistency

webknjaz commented on tests/github_app_test.py:
```diff
@@ -271,16 +177,16 @@ def token(self: '_FakeAppInstallationAuth') -> str:
+        'Iv2aaaaaaaaaaaaaaaa',
+        'Iv10aaaaaaaaaaaaaaaa',
+        'Iv100.aaaaaaaaaaaaaaaa',
+        'Iv23likIfIXeZTb5GCAA',
```
These params become a part of the test ID. They show up in test reports and console output. It's not necessarily clear what they are for.
This is why, please wrap each entry with `pytest.param(..., id='human-oriented-test-case-identifier')` which would make it clear why these are being tested in the first place.
Don't use the separate `ids=` argument to `pytest.mark.parametrize()` since the list of things is too long for that style to be maintainable in this specific case.

stevensonmichel committed:
Add descriptive test IDs for parametrized test cases

The test parameters are now wrapped with pytest.param() to provide
human-readable test identifiers. This makes test output and reports
much clearer by showing descriptive names like 'client-id-iv23-real-
example' instead of raw values like 'Iv23likIfIXeZTb5GCAA'.

This improves test debugging and makes it immediately obvious what
each test case is validating when viewing test results in CI or
console output.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

arrestle reviewed (approved):
Hi Stevenson! Good job getting this reviewed it's a tough one. 

I approve of the approach but honestly don't completely understand the implications of changing application_id to  application_or_client_id. However, I trust your team has enough integration testing to make sure this all works correctly! 

Let me know if you need any further assistance.

stevensonmichel commented:
Thank you so much for the review and approval! I really appreciate your feedback. Regarding the application_id to application_or_client_id change, the name was updated to better reflect that it now accepts both GitHub App IDs and Client IDs. The key change in this PR is letting PyGitHub's API be the ultimate validator. This makes the plugin future-proof when GitHub introduces new ID formats.
